### PR TITLE
HPCC-16630 Add expiry parameter to spray commands

### PR DESCRIPTION
**Pull request by AttilaVamos**
Add 'expireDays' parameter to Spray, in Std.File ECL code and DfuPlus CLI.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@ghalliday please review.

File: ecllibrary/std/File.ecl [hunk 1]
```diff
@@ -412,27 +412,28 @@ EXPORT DfuPlusExec(varstring cmdline) :=
  * @param replicate     Whether to replicate the new file. Defaults to FALSE.
  * @param compress      Whether to compress the new file. Defaults to FALSE.
  * @param failIfNoSourceFile If TRUE it causes a missing source file to trigger a failure.  Defaults to FALSE.
+ * @param expireDays    Number of dayt to auto-remove file. Default is '' (empty), not expire.
```
Comment (ghalliday) [inline ecllibrary/std/File.ecl]: tyop dayt->days

**Review by ghalliday — COMMENTED** [review]
It is a bit strange for the expireDays parameter to be a string though.  It would be more natural for it to be an integer (with a default -ve value to indicate it never expires).

Otherwise the changes look good.

**Comment by ghalliday** [comment]
@AttilaVamos one comment, 1 typo (repeated a few times)

**Comment by AttilaVamos** [comment]
@Gavin typo fixed.

**Comment by richardkchapman** [comment]
@ghalliday This is for you

**Comment by ghalliday** [comment]
@richardkchapman what do you think about the expireDays parameter being a string?  Otherwise ready to merge.

**Comment by richardkchapman** [comment]
It should be an integer. Can use 0 and/or -ve to signify no expiry

**Comment by richardkchapman** [comment]
@AttilaVamos Was there a reason for using a string?

**Comment by ghalliday** [comment]
@AttilaVamos back to you.

**Comment by AttilaVamos** [comment]
@richardkchapman The reasons are:
- It never used that value in any calculation during spray
- It stored (like any other values) as a string into file properties
(- furthermore I can imagine a later version where the value if expire can be 1w for one week or 2m for 2 months, etc.)
However if you or @ghalliday thinks it would be better as a numeric then I can modify my implementation.

**Comment by ghalliday** [comment]
@AttilaVamos from the ECL programmers point of view the "expireDays" is a numeric quantity (expirePeriod would possibly suggest a string).  Similarly the parameter to the existing EXPIRE() attribute is also a numeric.
Most of the code will stay the same - it is only the fileservices functions, and the esp interface.

I think it needs to be a -ve number for the default (in the fileservices functions) since 0 means expire, but use the default period.

**Comment by AttilaVamos** [comment]
OK,  I will update my code soon.

**Comment by AttilaVamos** [comment]
@ghalliday Reworked, please re-review.

**Comment by AttilaVamos** [comment]
@ghalliday something went wrong. I'm checking

**Comment by ghalliday** [comment]
ok @AttilaVamos

**Comment by AttilaVamos** [comment]
@ghalliday reworked, mistake fixed and regression test case added. Please re-review.

File: dali/dfu/dfuwu.cpp [hunk 2]
```diff
@@ -2077,6 +2077,16 @@ class CDFUoptions: public CLinkedDFUWUchild, implements IDFUoptions
         queryRoot()->setProp("@umask",val);
 
     }
+    int getExpireDays() const
+    {
+        if (queryRoot()->hasProp("@expireDays"))
```
Comment (ghalliday) [inline dali/dfu/dfuwu.cpp]: minor: No need to check this condition - since it will return -1 if the attribute doesn't exist.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**